lib/twirp/rails/logging/adapter.rb

```diff
@@ -6,19 +6,20 @@ module Twirp
   module Rails
     module Logging
       module Adapter# :nodoc:
+
         def self.install
           return unless defined?(ActiveSupport::Notifications)
 
-          Twirp::Rails::Routes.on_create_service do |service|
-            Logging::Adapter.instrument service
+          Twirp::Rails::Routes.on_create_service do |service_wrapper|
+            Logging::Adapter.instrument service_wrapper
           end
         end
 
-        def self.instrument(service)
+        def self.instrument(service_wrapper)
           instrumenter = ActiveSupport::Notifications.instrumenter
 
-          service.before do |rack_env, env|
-            env = env.merge(service: service)
+          service_wrapper.before_route_request do |rack_env, env|
+            env[:service] = service_wrapper.service
```

Comment:
why do we call `env[:service] = service_wrapper.service` 3 times in 3 different blocks when, i think, `env` is always the same and `service_wrapper` is always the same? 

maybe this is a 6th method?

Reply:
@masumoki Because this is a wrapper around `Twirp::Service`, the `before_route_request` is a non-standard hook that is injected before the `Twirp::Service#call` method. Therefore the `env` in `before_route_request` has not yet been built as that happens within `Twirp::Service#call` after a potential `:bad_route` or `:malformed` error code. Therefore the `env` gets reinstantiated immediately after `before_route_request`. To remedy this, we have to re-inject the service ASAP in the `before` hook. However, the `error` hook could trigger after `before_route_request`, after `env` is reinstantiated, but before `before` hook. This leaves `env` without `service`